packages/notifier/src/publish-kit.js

```diff
@@ -216,75 +216,90 @@ const initDurablePublishKitState = (options = {}) => {
  */
 const getEphemeralKey = facets => facets.publisher;
 
-/** @type {WeakMap<DurablePublishKitEphemeralKey, {currentP, tailP, tailR}>} */
+/**
+ * @typedef DurablePublishKitEphemeralData
+ * @property {Promise<*> | undefined} currentP The current-result promise
+ *   (undefined unless resolved with unrecoverable ephemeral data)
+ * @property {Promise<*>} tailP The next-result promise
+ * @property {import('@endo/promise-kit').PromiseKit<*>['resolve'] | undefined} tailR The next-result resolver
+ *   (undefined when the publish kit has terminated)
+ */
+
+/** @type {WeakMap<DurablePublishKitEphemeralKey, DurablePublishKitEphemeralData>} */
 const durablePublishKitEphemeralData = new WeakMap();
 
 /**
- * Returns the current/next-result promises and next-result resolver
- * associated with a given durable publish kit.
- * They are lost on upgrade, but recreated on-demand.
- * Such recreation preserves the value in (but not the identity of) the
- * current { value, done } result when possible, which is always the
- * case when that value is terminal (i.e., from `finish` or `fail`) or
- * when the durable publish kit is configured with
- * `valueDurability: 'mandatory'`.
+ * Returns the current-result promise associated with a given durable
+ * publish kit, recreated unless we already have one with retained
+ * ephemeral data.
  *
  * @param {DurablePublishKitState} state
  * @param {PublishKit<*>} facets
+ * @param {Promise<*>} tail
+ * @returns {Promise<*>}
+ */
+const provideCurrentP = (state, facets, tail) => {
+  const ephemeralKey = getEphemeralKey(facets);
+  const foundData = durablePublishKitEphemeralData.get(ephemeralKey);
+  const currentP = foundData?.currentP;
+  if (currentP) {
+    return currentP;
+  }
+
+  const { publishCount, status, hasValue, value } = state;
+  if (!hasValue) {
+    assert(status === 'live');
+    return tail;
+  }
+  if (status === 'live' || status === 'finished') {
```

Comment:
Maybe use `switch (status)` here?

Reply:
Considered, but rejected because I want `provideCurrentP` and `provideDurablePublishKitEphemeralData` to remain compact.